libshpool/src/config_watcher.rs

```diff
@@ -0,0 +1,720 @@
+            let (watched_path, immediate_child) =
+                best_effort_watch(&mut watcher, &target_path).unwrap();
+
+            assert_eq!(watched_path, target_path);
+            assert_eq!(immediate_child, None);
+        }
+    }
+
+    mod handle_event {
+        use super::*;
+        use assert_matches::assert_matches;
+        use notify::{
+            event::{CreateKind, ModifyKind, RemoveKind, RenameMode},
+            Event, EventKind,
+        };
+        use ntest::test_case;
+
+        fn paths_entry(target: &str, watched: &str) -> (PathBuf, (PathBuf, PathBuf)) {
+            let target = PathBuf::from(target);
+            let base = PathBuf::from(watched);
+            let immediate =
+                base.join(target.strip_prefix(&base).unwrap().iter().next().unwrap_or_default());
+            (target, (base, immediate))
+        }
+
+        // create event from spec:
+        // <create|mv|modify|rm> path1 [path2]
+        // `base` is prepended to all paths
+        fn event_from_spec(base: &str, evt: &str) -> notify::Event {
+            let base = Path::new(base);
+            let (evt, path) = evt.split_once(' ').unwrap_or((evt, ""));
+            match evt {
+                "create" => {
+                    Event::new(EventKind::Create(CreateKind::Any)).add_path(base.join(path))
+                }
+                "mv" => {
+                    let (src, dst) = path.split_once(' ').unwrap();
+                    Event::new(EventKind::Modify(ModifyKind::Name(RenameMode::Both)))
+                        .add_path(base.join(src))
+                        .add_path(base.join(dst))
+                }
+                "mvselfother" => Event::new(EventKind::Modify(ModifyKind::Name(RenameMode::Both)))
+                    .add_path(base.to_owned())
+                    .add_path(PathBuf::from("/some/other/path")),
+                "modify" => {
+                    Event::new(EventKind::Modify(ModifyKind::Any)).add_path(base.join(path))
+                }
+                "modifyself" => {
+                    Event::new(EventKind::Modify(ModifyKind::Any)).add_path(base.to_owned())
+                }
+                "rm" => Event::new(EventKind::Remove(RemoveKind::Any)).add_path(base.join(path)),
+                "rmself" => {
+                    Event::new(EventKind::Remove(RemoveKind::Any)).add_path(base.to_owned())
+                }
+                _ => panic!("malformatted event spec"),
+            }
+        }
+
+        #[test]
+        fn need_rescan() {
+            let event = notify::Event::default().set_flag(notify::event::Flag::Rescan);
+            let paths = Default::default();
+            let (rewatch, reload) = handle_event(event, &paths);
+            assert_eq!(rewatch, ReWatch::All);
+            assert!(reload);
+        }
+
+        const TARGET: &str = "/base/sub/config.toml";
+
+        #[test_case(TARGET, "/base", "create sub", true, false, name = "base_create_sub")]
```

Comment:
I just tried using the annotation in another change I'm working on and got errors about this being an unstable feature. We should keep shpool building and working on stable rust (and try to avoid bumping the MSRV so that lagging distros can start picking it up once they catch up to the current MSRV).

Reply:
Ha this also confused me when trying this for the first time. 

This is a macro from ntest and you need to import it. Otherwise, the macro from std of the same name is picked up, which is unstable.

And +1 on maintaining MSRV.

Reply:
Oh, I see, thanks